network/requestTracker.go

```diff
@@ -0,0 +1,385 @@
+// Copyright (C) 2019 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"sort"
+	"time"
+
+	"github.com/algorand/go-deadlock"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/logging/telemetryspec"
+)
+
+const (
+	// maxHeaderReadTimeout is the time limit where items would remain in the acceptedConnections cache before being pruned.
+	// certain malicious connections would never get to the http handler, and therefore must be pruned every so often.
+	maxHeaderReadTimeout = 30 * time.Second
```

Comment:
should the maxHeaderReadTimeout be configurable?

Reply:
not really. If we "fail" with this timeout, it just mean that the passed TrackerRequest structure would be generated on our http handler. This is sub-optimal, because we loose track of the original TrackerRequest creation time, but has no practical negative implications otherwise.